// gumaddon/src/Elements/AddonEntityPack.js
class AddonEntityPack {
    #clientfile
    #serverfile
    #geometries
    #textures
    #rendercontrollers
    #animations
    #animationcontrollers

    /** @param {String} typeId */
    constructor(typeId) {
        this.typeId = typeId;
        this.pack_name = `[pack]${typeId}`;
        
        this.#clientfile = {};
        this.#serverfile = {};
        this.#geometries = [];
        this.#textures = {};
        this.#rendercontrollers = {};
        this.#animations = {};
        this.#animationcontrollers = {};
    }

    /** @param {String} content */
    setClientFile( content ){
        this.#clientfile = JSON.parse( content );
    }

    /** @param {String} content */
    setServerFile( content ){
        this.#serverfile = JSON.parse( content );
    }

    /** 
     * @param {String} name
     * @param {String} content
     */
    addGeometryFile(content) {
        this.#geometries.push(JSON.parse(content));
    }

    /** 
     * @param {String} name
     * @param {String} rpath
     */
    addTexture(name, rpath) {
        this.#textures[name] = rpath;
    }

    /** 
     * @param {String} identifier
     * @param {String} content
     */
    addRenderController(identifier, content) {
        this.#rendercontrollers[identifier] = JSON.parse(content);
    }
}

module.exports = { AddonEntityPack };
